pgpool/kubedb_client_builder.go

```diff
@@ -0,0 +1,180 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pgpool
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	_ "github.com/lib/pq"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	appbinding "kmodules.xyz/custom-resources/apis/appcatalog/v1alpha1"
+	api "kubedb.dev/apimachinery/apis/kubedb/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"xorm.io/xorm"
+)
+
+const (
+	DefaultPgpoolDB = "postgres"
+)
+
+type KubeDBClientBuilder struct {
+	kc       client.Client
+	db       *api.Pgpool
+	url      string
+	podName  string
+	pgpoolDB string
+	ctx      context.Context
+}
+
+func NewKubeDBClientBuilder(kc client.Client, db *api.Pgpool) *KubeDBClientBuilder {
+	return &KubeDBClientBuilder{
+		kc: kc,
+		db: db,
+	}
+}
+
+func (o *KubeDBClientBuilder) WithURL(url string) *KubeDBClientBuilder {
+	o.url = url
+	return o
+}
+
+func (o *KubeDBClientBuilder) WithPod(podName string) *KubeDBClientBuilder {
+	o.podName = podName
+	return o
+}
+
+func (o *KubeDBClientBuilder) WithPgpoolDB(pgDB string) *KubeDBClientBuilder {
+	o.pgpoolDB = pgDB
+	return o
+}
+
+func (o *KubeDBClientBuilder) WithContext(ctx context.Context) *KubeDBClientBuilder {
+	o.ctx = ctx
+	return o
+}
+
+func (o *KubeDBClientBuilder) GetPgpoolXormClient() (*XormClient, error) {
+	if o.ctx == nil {
+		o.ctx = context.Background()
+	}
+
+	connector, err := o.getConnectionString()
+	if err != nil {
+		return nil, err
+	}
+
+	engine, err := xorm.NewEngine("postgres", connector)
+	if err != nil {
+		return nil, err
+	}
+	_, err = engine.Query("SELECT 1")
+	if err != nil {
+		return nil, err
+	}
+
+	engine.SetDefaultContext(o.ctx)
+	return &XormClient{
+		engine,
+	}, nil
+}
+
+func (o *KubeDBClientBuilder) GetPgpoolClient() (*Client, error) {
+	if o.ctx == nil {
+		o.ctx = context.Background()
+	}
+
+	connector, err := o.getConnectionString()
+	if err != nil {
+		return nil, err
+	}
+
+	// connect to database
+	db, err := sql.Open("postgres", connector)
+	if err != nil {
+		return nil, err
+	}
+
+	// ping to database to check the connection
+	if err := db.PingContext(o.ctx); err != nil {
+		closeErr := db.Close()
+		if closeErr != nil {
+			klog.Errorf("Failed to close client. error: %v", closeErr)
+		}
+		return nil, err
+	}
+
+	return &Client{db}, nil
+}
+
+func (o *KubeDBClientBuilder) getURL() string {
+	return fmt.Sprintf("%s.%s.%s.svc", o.podName, o.db.GoverningServiceName(), o.db.Namespace)
+}
+
+func (o *KubeDBClientBuilder) getPgpoolAuthCredentials() (string, string, error) {
+	db := o.db
+	var secretName string
+	if db.Spec.Backend != nil {
+		apb := &appbinding.AppBinding{}
+		err := o.kc.Get(o.ctx, types.NamespacedName{
+			Name:      db.Spec.Backend.Name,
+			Namespace: db.Namespace,
+		}, apb)
+		if err != nil {
+			return "", "", err
+		}
+		if apb.Spec.Secret == nil {
+			return "", "", fmt.Errorf("No database secret")
```

Comment:
change "no database secret" to "backend database auth secret not found" .or something better